### DISCUSSION TITLE
FTP Server for nanoFramework as a way to move files to and from MCU

### DISCUSSION THREAD

**Discussion by icy3141** [nanoframework/Home#1572] [General]
### Description

- There does not appear to be a straightforward way to move files to and from the internal storage of an MCU device running nanoFramework.
- Various projects may benefit from FTP technology, regardless of other ways of moving files. (i.e. gathering data from sensors, updating configuration files or web server files, etc.)

### How to solve the problem

An implementation of a basic FTP server usable in nanoFramework projects could provide a fairly straightforward way to move files to and from an MCU device running nanoFramework.

### Describe alternatives you've considered

1. Deploy files as Resources
So far, this is how I have been moving files to my microcontroller to use for a web interface. It is a somewhat cumbersome solution because it requires deploying a separate project to the MCU before deploying the main application, and that secondary project has to be updated anytime there are changes to the file structure that's meant to be moved over.
2. Write a file server application using nanoFramework and a companion client-side app.
There are many implementations of things like this for Arduino that could be used as a guide, but I would rather avoid developing a client-side app for this, and Filezilla can already serve as that with FTP.

### Aditional context

There is a free guide on implementing FTP server in C# with source code using the MIT license. Using this as a foundation, implementing a barebones FTP server could be relatively simple.
https://www.codeproject.com/articles/380769/creating-an-ftp-server-in-csharp-with-ipv6-support

Given my limitations, I want to avoid committing to handling everything to do with this in its entirety on my own, but I could at least get some basic functionality working and submit it.

I have started on this recently and yesterday I got a file to move over to my ESP32 and back to my PC.

I was thinking also there could be a SPIFFS representation of folders using special characters in the filenames (configurable from the code instantiating the server), so that directories could be moved back and forth without an issue.

**Comment by Ellerbach** [comment]
Moving this as a discussion as we keep issues for issues :-)
.NET nanoFramework is primarily a community driven project, so, if someone from the community wants a feature, we, as the maintainers, are always happy to support.
If you are motivated to implement FTP, please go ahead! Let us know on Discord if you need any assistance.

Now, for deploying at build time, you have the Visual Studio extension allowing you to mark files as "Content" and anything else than "Never" will deploy those files in the internal storage.
And nanoff allows you as well to deploy files at flash time!
For security reasons, we're limited how to get files back from those storage to the PC. Not that it is impossible, just that we prefer to treat the files as potential secrets. As many storage are not really secure, we prefer to limit getting them.
Today, if you want to download files that are not markets as system, hidden or .SYS files, you can write a simple web server listing the entire file structure and serving you the files.
